pkgs/build-support/portable-service/default.nix

```diff
@@ -0,0 +1,118 @@
+{ pkgs, lib, stdenv }:
+/*
+  Create a systemd portable service image
+  https://systemd.io/PORTABLE_SERVICES/
+
+  Example:
+  pkgs.portableService {
+      name = "demo";
+      units = [ demo-service demo-socket ];
+  }
+*/
+{
+  # The name and version of the portable service. The resulting image will be
+  # created in result/$name_$version.raw
+  name
+, version ? "dev"
+
+  # Units is a list of derivations for systemd unit files. Those files will be
+  # copied to /etc/systemd/system in the resulting image. Note that the unit
+  # names must be prefixed with the name of the portable service.
+, units
+
+  # Basic info about the portable service image, used for the generated
+  # /etc/os-release
+, description ? null
+, homepage ? null
+
+  # A list of attribute sets {object, symlink}. Symlinks will be created
+  # in the root filesystem of the image to objects in the nix store.
+, symlinks ? [ ]
+
+  # A list of additional derivations to be included in the image as-is.
+, contents ? [ ]
+
+  # mksquashfs options
+, squashfsTools ? pkgs.squashfsTools
+, squash-compression ? "xz -Xdict-size 100%"
+, squash-block-size ? "1M"
+}:
+
+let
+  filterNull = lib.filterAttrs (_: v: v != null);
+  envFileGenerator = lib.generators.toKeyValue { };
+
+  rootFsScaffold =
+    let
+      os-release-params = {
+        PORTABLE_ID = name;
+        PORTABLE_PRETTY_NAME = description;
+        HOME_URL = homepage;
+        ID = "nixos";
+        PRETTY_NAME = "NixOS";
+        BUILD_ID = "rolling";
+      };
+      os-release = pkgs.writeText "os-release"
+        (envFileGenerator (filterNull os-release-params));
+
+    in
+    stdenv.mkDerivation {
+      name = "root-fs-scaffold";
+
+      symlinkNames = map (x: x.symlink) symlinks;
+      symlinkTargets = map (x: x.object) symlinks;
+
+      buildCommand = ''
+        # scaffold a file system layout
+        mkdir -p $out/etc/systemd/system $out/proc $out/sys $out/dev $out/run $out/tmp $out/var/tmp
+
+        # empty files to mount over with host's version
+        touch $out/etc/resolv.conf $out/etc/machine-id
+
+        # required for portable services
+        cp ${os-release} $out/etc/os-release
+
+        # units **must** be copied to /etc/systemd/system/
+        ${lib.concatMapStringsSep "\n" (u: "cp ${u} $out/etc/systemd/system/${u.name};") units}
+
+        # Add symlinks to the top-level store objects.
+        symlinkTargets=($symlinkTargets)
+        symlinkNames=($symlinkNames)
+        for ((n = 0; n < ''${#symlinkTargets[*]}; n++)); do
+          target=''${symlinkTargets[$n]}
+          symlink=''${symlinkNames[$n]}
+          if test "$symlink" != "none"; then
+              mkdir -p $(dirname "$out/$symlink")
+              ln -sT "$target" "$out/$symlink"
+          fi
+        done
```

Comment:
I would recommend to use nix code to generate this rather than relying on respliting bash variables.

Reply:
indeed, I copied that from some other .nix files. Let me rethink that

Reply:
something like this 340257ba5ec3094eff062c0439f136483e97eff7 ?

Reply:
I just took a brief look but that already looks good